contracts/airdrop/AirdropRewarder.sol

```diff
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.20;
+
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {ERC20Upgradeable} from "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import {MerkleProof} from "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+contract AirdropRewarder is Initializable, OwnableUpgradeable {
+    using SafeERC20 for ERC20Upgradeable;
+    address public safeAddress;
+
+    mapping(address => bytes32) private merkleRoots;
+    mapping(address => mapping(address => uint256)) private totalRewardsClaimed;
+
+    error InvalidAddress();
+    error InvalidMerkleProof(bytes32[]);
+
+    event MerkleRootSetForToken(bytes32 _merkleRoot, address _rewardToken);
+    event SafeAddressSet(address _safeAddress);
+    event RewardsClaimed(
+        address _user,
+        address _rewardsToken,
+        uint256 _rewardsAmount
+    );
+    event RewardTerminated(address _rewardAddress);
+
+    function initialize(address _safeAddress) external initializer {
+        __Ownable_init(msg.sender);
+        safeAddress = _safeAddress;
+        emit SafeAddressSet(_safeAddress);
+    }
+
+    function setMerkleRoot(
+        bytes32 _merkleRoot,
+        address _rewardToken
+    ) external onlyOwner {
+        merkleRoots[_rewardToken] = _merkleRoot;
+        emit MerkleRootSetForToken(_merkleRoot, _rewardToken);
+    }
+
+    function claim(
+        address _user,
+        address rewardToken,
+        uint256 _totalClaimableAmount,
+        bytes32[] calldata _merkleProofs
+    ) external {
+        uint256 totalClaimableAmount = _totalClaimableAmount;
+        address rewardAddress = rewardToken;
+        if (rewardAddress == address(0)) revert InvalidAddress();
+
```

Comment:
need to also write a check to make sure a user has not already claimed the airdrop.

Reply:
Done